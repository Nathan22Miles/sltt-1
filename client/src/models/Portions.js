// Portion actions, unit test

import { types, getParent } from "mobx-state-tree"
import _ from 'underscore'

import { Passage } from './Passages.js'
import { _insertBy, _remove } from './ModelUtils.js'


function _move(db, items, _id, newIndex, cb) { 
    let oldIndex = _.findIndex(items, { _id })
    if (oldIndex < 0) {
        cb(`_move: cannot find ${_id}`)
        return
    }

    if (oldIndex === newIndex) return

    // Moving an array element which is a model from array element oldIndex to newIndex is tricky. 
    // As far as I can tell you can't just use splice
    // to extract and then reinsert a model in mobx-state-tree array.
    // Removing the item from the array 'kills' it and then it cannot be resinerted.
    // Instead you need to get a plain js object representing the model and then reinsert that.
    let item = items[oldIndex].toJSON()
    items.splice(oldIndex, 1)
    items.splice(newIndex, 0, item)

    let prevRank = 0.0
    let nextRank = 200.0

    if (newIndex > 0) {
        prevRank = items[newIndex-1].rank
        nextRank = prevRank + 200.0
    }
    if (newIndex < items.length-1) { 
        nextRank = items[newIndex+1].rank
    }
    let newRank = (prevRank + nextRank) / 2.0
    items[newIndex].rank = newRank

    db.get(_id)
        .then(doc => {
            doc.rank = newRank
            return db.put(doc)
        })
        .then(() => cb && cb())
        .catch(err => { cb && cb(err) })
}

function _apply(doc, list) {
    if (doc._deleted) {
        let idx = _.findIndex(list, { _id: doc._id })
        if (idx < 0) return
        list.splice(idx, 1)
        return
    }

    _insertBy(doc, list, 'rank')
}

function _add(_db, doc, list, cb) {
    let rank = 100.0
    if (list.length > 0) rank = list[list.length - 1].rank + 100.0
    doc.rank = rank

    list.push(doc)

    _db.put(doc)
        .then(() => cb && cb())
        .catch(err => { 
            console.log(err)
            cb && cb(err) 
        })
}

function _invalidNameCheck(name) {
    let match = name.match(/^[\w \-.]+$/)
    if (!match) return 'Must have only letters, numbers, hyphen(-), periods(.), and spaces'
    return null
}

function _duplicateCheck(items, name) {
    name = name.trim()
    if (_.findWhere(items, { name }))
        return 'Duplicate name'

    return null
}


// ================== PORTION =====================

export const Portion = types.model("Portion", {
    _id: types.identifier(),
    name: types.string,
    rank: types.number,
    passages: types.array(Passage, [])
})
.actions(self => ({
    load: () => {
        let options = {
            startkey: self.name + '/',
            endkey: self.name + '/ufff0',
            include_docs: true,
        }

        return new Promise((resolve, reject) => {
            //console.log("AAA", self.getProject().name)

            self.getDb()
                .allDocs(options)
                .then(response => {
                    // response = {rows: [{doc: {...} }]}
                    for (let row of response.rows) {
                        self.apply(row.doc)
                    }
                    //console.log("BBB", self.getProject().name)
                    resolve()
                })
                .catch(err => {
                    let message = `[${self.getProject().name}/${self.name}] Portion.load ${err.stack}`
                    console.error(message)
                    reject(message)
                })
        })
    },

    getDb() {
        let portions = getParent(self, 2)
        let _db = portions.getDb()
        return _db
    },

    apply: (doc) => {
        let parts = doc._id.split('/')
        
        if (parts[0] === '#item') {
            if (!doc.statuses) doc.statuses = [] // data migration
            _apply(doc, self.passages)
            return
        }

        let name = parts[1]
        let passage = _.findWhere(self.passages, { name })
        if (!passage) {
            console.error(`no matching passage for ${doc._id}`)
            //throw new Error('') //!!
            return
        }

        passage.apply(doc)
    },

    movePassage: (_id, newIndex, cb) => { 
        let _db = self.getDb()
        _move(_db, self.passages, _id, newIndex, cb) 
    },    

    removePassage: (_id, cb) => { 
        let _db = self.getDb()
        _remove(_db, self.passages, _id, cb) 
    },    

    addPassage: (name, cb) => {
        let err = self.checkName(name)
        if (err) {
            cb && cb(err)
            return
        }

        let _id = `#item/${self.name}/${name}/passage`
        let doc = { 
            _id, 
            name, 
            statuses: [],
            notes: [],
            videos: [],
        }
        _add(self.getDb(), doc, self.passages, cb)
    },

    checkName: (name) => {
        return _invalidNameCheck(name) || _duplicateCheck(self.passages, name)
    },

    getProject: () => {
        let project = getParent(self, 3)
        return project
    },

}))

// Portion.checkEq = function (por1, por2) {
//     let attrs = ['_id', 'name', 'rank']
//     return _.all(attrs, attr => por1[attr] === por2[attr])
// }


// ================== PORTIONS =======================

export const Portions = types.model("Portions", {
    portions: types.array(Portion, []),
})
.actions(self => {
    let _db

    return {
        load: () => {
            let options = {
                startkey: '#item/',
                endkey: '#item/ufff0',
                include_docs: true,
            }

            let isPortion = row => (row.doc._id.endsWith('/portion'))
            let isPassage = row => (row.doc._id.endsWith('/passage'))

            return new Promise((resolve, reject) => {
                _db.allDocs(options)
                    .then(response => {
                        // response = {rows: [{doc: {...} }]}

                        // Process portions first so we have an owner for each passage
                        _.chain(response.rows)
                            .filter(isPortion)
                            .forEach(row => self.apply(row.doc))
                            
                        _.chain(response.rows)
                            .filter(isPassage)
                            .forEach(row => self.apply(row.doc))

                        resolve()
                    })
                    .catch(err => reject(err))
            }) 
        },

        apply: (doc) => {
            //console.log('portions apply', doc._id)

            let parts = doc._id.split('/')
            let tag = parts[parts.length-1]

            if (parts[0] === '#item' && tag === 'portion') { 
                _apply(doc, self.portions)
                return
            }

            if (parts[0] === '#item' && tag === 'passage') {
                let portionName = doc._id.split('/')[1]
                
                // We inserted the portions first so we should 'always' find a matching 
                // portion for this passage
                let idx = _.findIndex(self.portions, { name: portionName })
                if (idx < 0) throw new Error(`Portion not found [${doc._id}]`)

                self.portions[idx].apply(doc)
                return
            }

            let name = parts[0]
            let portion = _.findWhere(self.portions, { name })
            if (!portion) {
                //!!! error
                return
            }

            portion.apply(doc)
        },

        setDb: (db) => { _db = db },

        getDb: () => { return _db },

        movePortion: (_id, newIndex, cb) => {
            _move(_db, self.portions, _id, newIndex, cb)
        },

        removePortion: (_id, cb) => {
            _remove(_db, self.portions, _id, cb)
        },

        addPortion: (name, cb) => {
            let err = self.checkName(name)
            if (err) {
                cb && cb(err)
                return
            }

            let _id = `#item/${name}/portion`
            let doc = { _id, name, passages: [] }
            _add(_db, doc, self.portions, cb)
        },

        checkName: (name) => {
            return _invalidNameCheck(name) || _duplicateCheck(self.portions, name)
        }, 

    }
})
// .views(self => ({
// }))

