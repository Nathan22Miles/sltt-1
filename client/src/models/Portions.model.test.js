import _ from 'underscore'
import PouchDB from 'pouchdb'

import { Portions, Portion } from './Portions.js'

// Create in-memory test DB
PouchDB.plugin(require('pouchdb-adapter-memory'))
let dbName = 'DB' + Math.random().toString(8)
let _test1 = new PouchDB(dbName, { adapter: 'memory' })


let portions, portion

it("Can create empty Portions", (done) => {
    portions = Portions.create({ portions: [] })
    portions.setDb(_test1)
    expect(portions.portions.length).toBe(0)
    done()
})

it("Can add portions 1 of 3", (done) => {
    portions.addPortion('Matthew', err => {
        expect(err).toBeFalsy()
        expect(portions.portions.length).toBe(1)
        expect(portions.portions[0].name).toBe('Matthew')
        done()
    })
})

it("Can add portions 2 of 3", (done) => {
    portions.addPortion('Luke', err => {
        expect(err).toBeFalsy()
        expect(portions.portions.length).toBe(2)
        expect(portions.portions[1].name).toBe('Luke')
        done()
    })
})

it("Can add portions 3 of 3", (done) => {
    portions.addPortion('Mark', err => {
        expect(err).toBeFalsy()
        expect(portions.portions.length).toBe(3)
        expect(portions.portions[2].name).toBe('Mark')
        done()
    })
})

it("Does not add duplicate portion", (done) => {
    portions.addPortion('Mark', err => {
        expect(err).toBe('Duplicate name')
        expect(portions.portions.length).toBe(3)
        done()
    })
})

it("Does not add portion with invalid name", (done) => {
    portions.addPortion('#Bob', err => {
        expect(err).toBeTruthy()
        expect(portions.portions.length).toBe(3)
        done()
    })
})


it("Can move portion", (done) => {
    let _id = portions.portions[2]._id
    portions.movePortion(_id, 1, err => {
        expect(err).toBeFalsy()
        let names = _.pluck(portions.portions, 'name')

        expect(names).toEqual(['Matthew', 'Mark', 'Luke'])
        done()
    })
})

it("Can remove portion", (done) => {
    let _id = portions.portions[0]._id
    portions.removePortion(_id, err => {
        expect(err).toBeFalsy()
        let names = _.pluck(portions.portions, 'name')

        expect(names).toEqual(['Mark', 'Luke'])
        done()
    })
})

// ========== Passages tests ============


it("Can add passages 1 of 3", (done) => {
    portion = portions.portions[0]
    let name = 'Mark 1.1-3'
    portion.addPassage(name, err => {
        expect(err).toBeFalsy()
        expect(portion.passages.length).toBe(1)
        expect(portion.passages[0].name).toBe(name)
        done()
    })
})

it("Can add passages 2 of 3", (done) => {
    let name = 'Mark 1.7-9'
    portion.addPassage(name, err => {
        expect(err).toBeFalsy()
        expect(portion.passages.length).toBe(2)
        expect(portion.passages[1].name).toBe(name)
        done()
    })
})

it("Can add passages 3 of 3", (done) => {
    let name = 'Mark 1.4-6'
    portion.addPassage(name, err => {
        expect(err).toBeFalsy()
        expect(portion.passages.length).toBe(3)
        expect(portion.passages[2].name).toBe(name)
        done()
    })
})

it("Does not add duplicate passage", (done) => {
    let name = 'Mark 1.1-3'
    portion.addPassage(name, err => {
        expect(err).toBeTruthy()
        expect(portion.passages.length).toBe(3)
        done()
    })
})

it("Does not add passage with invalid name", (done) => {
    portion.addPassage('#Bob', err => {
        expect(err).toBeTruthy()
        expect(portion.passages.length).toBe(3)
        done()
    })
})

it("Can move passage", (done) => {
    let _id = portion.passages[2]._id
    let names1 = _.pluck(portion.passages, 'name')

    portion.movePassage(_id, 1, err => {
        expect(err).toBeFalsy()
        let names = _.pluck(portion.passages, 'name')

        expect(names).toEqual(['Mark 1.1-3', 'Mark 1.4-6', 'Mark 1.7-9'])
        done()
    })
})

it("Can remove passage", (done) => {
    let _id = portion.passages[0]._id

    let names = _.pluck(portion.passages, 'name')
    portion.removePassage(_id, err => {
        expect(err).toBeFalsy()
        names = _.pluck(portion.passages, 'name')

        expect(names).toEqual(['Mark 1.4-6', 'Mark 1.7-9'])
        done()
    })
})

it("Can load passages from DB", (done) => {
    let portions2 = Portions.create({ portions: [] })

    portions2.setDb(_test1)

    portions2.load()
        .then(() => {
            expect(portions2).toMatchSnapshot()
            done()
        })
        .catch(err => {
            expect(err).toBeFalsy()  // Fail expectation on purpose
        })
})
