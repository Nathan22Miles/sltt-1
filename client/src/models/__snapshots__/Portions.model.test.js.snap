// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can load passages from DB 1`] = `
Object {
  "portions": Array [
    Object {
      "_id": "#item/Mark/portion",
      "name": "Mark",
      "passages": Array [
        Object {
          "_id": "#item/Mark/Mark 1.4-6/passage",
          "name": "Mark 1.4-6",
          "notes": Array [],
          "rank": 150,
          "statuses": Array [],
          "videos": Array [],
        },
        Object {
          "_id": "#item/Mark/Mark 1.7-9/passage",
          "name": "Mark 1.7-9",
          "notes": Array [],
          "rank": 200,
          "statuses": Array [],
          "videos": Array [],
        },
      ],
      "rank": 150,
    },
    Object {
      "_id": "#item/Luke/portion",
      "name": "Luke",
      "passages": Array [],
      "rank": 200,
    },
  ],
}
`;
